File: generators/generator-botbuilder/generators/app/templates/core/dialogs/dateResolverDialog.ts
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

import { TimexProperty } from '@microsoft/recognizers-text-data-types-timex-expression';
import { DateTimePrompt, DateTimeResolution, DialogTurnResult, PromptValidatorContext, WaterfallDialog, WaterfallStepContext } from 'botbuilder-dialogs';
import { CancelAndHelpDialog } from './cancelAndHelpDialog';

const DATETIME_PROMPT = 'datetimePrompt';
const WATERFALL_DIALOG = 'waterfallDialog';

export class DateResolverDialog extends CancelAndHelpDialog {
    constructor(id: string) {
        super(id || 'dateResolverDialog');
        this.addDialog(new DateTimePrompt(DATETIME_PROMPT, this.dateTimePromptValidator.bind(this)))
            .addDialog(new WaterfallDialog(WATERFALL_DIALOG, [
                this.initialStep.bind(this),
                this.finalStep.bind(this),
            ]));

        this.initialDialogId = WATERFALL_DIALOG;
    }

    private async initialStep(stepContext: WaterfallStepContext): Promise<DialogTurnResult> {
        const timex = (stepContext.options as any).date;

        const promptMsg = 'On what date would you like to travel?';
        const repromptMsg = 'I\'m sorry, for best results, please enter your travel date including the month, day and year.';

        if (!timex) {
            // We were not given any date at all so prompt the user.
            return await stepContext.prompt(DATETIME_PROMPT,
                {
                    prompt: promptMsg,
                    retryPrompt: repromptMsg,
                });
        } else {
            // We have a Date we just need to check it is unambiguous.
            const timexProperty = new TimexProperty(timex);
            if (!timexProperty.types.has('definite')) {
                // This is essentially a "reprompt" of the data we were given up front.
                return await stepContext.prompt(DATETIME_PROMPT, { prompt: repromptMsg });
            } else {
                return await stepContext.next({ timex });
            }
        }
    }

    private async finalStep(stepContext: WaterfallStepContext): Promise<DialogTurnResult> {
        const timex = stepContext.result[0].timex;
        return await stepContext.endDialog(timex);
    }

    private async dateTimePromptValidator(promptContext: PromptValidatorContext<DateTimeResolution>): Promise<boolean> {
        if (promptContext.recognized.succeeded) {
            // This value will be a TIMEX. And we are only interested in a Date so grab the first result and drop the Time part.
            // TIMEX is a format that represents DateTime expressions that include some ambiguity. e.g. missing a Year.
            const timex = promptContext.recognized.value[0].timex.split('T')[0];

            // If this is a definite Date including year, month and day we are good otherwise reprompt.
            // A better solution might be to let the user know what part is actually missing.
            return new TimexProperty(timex).types.has('definite');
        } else {
            return false;
        }
    }
}

File: samples/java_springboot/81.skills-skilldialog/dialog-skill-bot/src/main/java/com/microsoft/bot/sample/dialogskillbot/dialogs/DateResolverDialog.java
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MT License.

package com.microsoft.bot.sample.dialogskillbot.dialogs;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.concurrent.CompletableFuture;

import com.microsoft.applicationinsights.core.dependencies.apachecommons.lang3.StringUtils;
import com.microsoft.bot.builder.MessageFactory;
import com.microsoft.bot.dialogs.DialogTurnResult;
import com.microsoft.bot.dialogs.WaterfallDialog;
import com.microsoft.bot.dialogs.WaterfallStep;
import com.microsoft.bot.dialogs.WaterfallStepContext;
import com.microsoft.bot.dialogs.prompts.DateTimePrompt;
import com.microsoft.bot.dialogs.prompts.DateTimeResolution;
import com.microsoft.bot.dialogs.prompts.PromptOptions;
import com.microsoft.bot.dialogs.prompts.PromptValidator;
import com.microsoft.bot.dialogs.prompts.PromptValidatorContext;
import com.microsoft.bot.schema.Activity;
import com.microsoft.bot.schema.InputHints;
import com.microsoft.recognizers.datatypes.timex.expression.Constants;
import com.microsoft.recognizers.datatypes.timex.expression.TimexProperty;

public class DateResolverDialog extends CancelAndHelpDialog {

    private final String PromptMsgText = "When would you like to travel?";
    private final String RepromptMsgText = "I'm sorry, to make your booking please enter a full travel date, including Day, Month, and Year.";

    public DateResolverDialog(String id) {
        super(!StringUtils.isAllBlank(id) ? id : "DateResolverDialog");
        addDialog(new DateTimePrompt("DateTimePrompt", new dateTimePromptValidator(), null));
        WaterfallStep[] waterfallSteps = { this::initialStep, this::finalStep };
        addDialog(new WaterfallDialog("WaterfallDialog", Arrays.asList(waterfallSteps)));

        // The initial child Dialog to run.
        setInitialDialogId("WaterfallDialog");
    }

    class dateTimePromptValidator implements PromptValidator<List<DateTimeResolution>> {

        @Override
        public CompletableFuture<Boolean> promptValidator(
                PromptValidatorContext<List<DateTimeResolution>> promptContext) {
            if (promptContext.getRecognized().getSucceeded()) {
                // This value will be a TMEX. We are only interested in the Date part, so grab
                // the first result and drop the Time part.
                // TMEX instanceof a format that represents DateTime expressions that include
                // some ambiguity, such as a missing Year.
                String timex = promptContext.getRecognized().getValue().get(0).getTimex().split("T")[0];

                // If this instanceof a definite Date that includes year, month and day we are
                // good; otherwise, reprompt.
                // A better solution might be to let the user know what part instanceof actually
                // missing.
                Boolean isDefinite = new TimexProperty(timex).getTypes().contains(Constants.TimexTypes.DEFINITE);

                return CompletableFuture.completedFuture(isDefinite);
            }
            return CompletableFuture.completedFuture(false);
        }

    }

    public CompletableFuture<DialogTurnResult> initialStep(WaterfallStepContext stepContext) {
        String timex = (String) stepContext.getOptions();

        Activity promptMessage = MessageFactory.text(PromptMsgText, PromptMsgText, InputHints.EXPECTING_INPUT);
        Activity repromptMessage = MessageFactory.text(RepromptMsgText, RepromptMsgText, InputHints.EXPECTING_INPUT);

        if (timex == null) {
            // We were not given any date at all so prompt the user.
            PromptOptions options = new PromptOptions();
            options.setPrompt(promptMessage);
            options.setRetryPrompt(repromptMessage);
            return stepContext.prompt("DateTimePrompt", options);
        }

        // We have a Date we just need to check it instanceof unambiguous.
        TimexProperty timexProperty = new TimexProperty(timex);
        if (!timexProperty.getTypes().contains(Constants.TimexTypes.DEFINITE)) {
            // This instanceof essentially a "reprompt" of the data we were given up front.
            PromptOptions options = new PromptOptions();
            options.setPrompt(repromptMessage);
            return stepContext.prompt("DateTimePrompt", options);
        }
        List<DateTimeResolution> resolutionList = new ArrayList<DateTimeResolution>();
        DateTimeResolution resolution = new DateTimeResolution();
        resolution.setTimex(timex);
        resolutionList.add(resolution);
        return stepContext.next(resolutionList);
    }

    private CompletableFuture<DialogTurnResult> finalStep(WaterfallStepContext stepContext) {
        String timex = ((List<DateTimeResolution>) stepContext.getResult()).get(0).getTimex();
        return stepContext.endDialog(timex);
    }
}
